assets/js/base/hooks/checkout/use-checkout-address.js

```diff
@@ -0,0 +1,107 @@
+/**
+ * External dependencies
+ */
+import defaultAddressFields from '@woocommerce/base-components/cart-checkout/address-form/default-address-fields';
+import { useState, useEffect } from '@wordpress/element';
+import {
+	useShippingDataContext,
+	useBillingDataContext,
+	useCheckoutContext,
+} from '@woocommerce/base-context';
+import isShallowEqual from '@wordpress/is-shallow-equal';
+
+/**
+ * Compare two addresses and see if they are the same.
+ *
+ * @param {Object} address1 First address.
+ * @param {Object} address2 Second address.
+ */
+const isSameAddress = ( address1, address2 ) => {
+	return Object.keys( defaultAddressFields ).every(
+		( field ) => address1[ field ] === address2[ field ]
+	);
+};
+
+/**
+ * Custom hook for tracking address field state on checkout and persisting it to
+ * context globally on change.
+ */
+export const useCheckoutAddress = () => {
+	const { customerId } = useCheckoutContext();
+	const {
+		shippingAddress,
+		setShippingAddress,
+		needsShipping,
+	} = useShippingDataContext();
+	const { billingData, setBillingData } = useBillingDataContext();
+
+	// These are the local states of address fields, which are persisted
+	// globally when changed. They default to the global shipping address which
+	// is populated from the current customer data or default location.
+	const [ shippingFields, updateShippingFields ] = useState(
+		shippingAddress
+	);
+	const [ billingFields, updateBillingFields ] = useState( billingData );
+
+	// This tracks the state of the "shipping as billing" address checkbox. It's
+	// initial value is true (if shipping is needed), however, if the user is
+	// logged in and they have a different billing address, we can toggle this off.
+	const [ shippingAsBilling, setShippingAsBilling ] = useState(
+		() =>
+			needsShipping &&
+			( ! customerId || isSameAddress( shippingAddress, billingData ) )
+	);
+
+	// Pushes to global state when changes are made locally.
+	useEffect( () => {
+		if ( ! isShallowEqual( shippingFields, shippingAddress ) ) {
+			setShippingAddress( shippingFields );
+		}
+
+		const newBillingData = shippingAsBilling
+			? shippingFields
+			: billingFields;
+
+		if ( ! isShallowEqual( newBillingData, billingData ) ) {
+			setBillingData( newBillingData );
+		}
```

Comment:
_State_ might not be; it can be an object. This should be switched to lodash isEqual perhaps.

Reply:
`isEqual` is slow, so if we can avoid it, that would be good because this is a fairly hot code path. Still, you could try it and see how it impacts things.

Reply:
@nerrad I was able to avoid it for the validation class in the end. But the address one I did use it, because of state and country and possibly other data where we need to compare the entire address object, not just a new/specific property.